File: sdk/src/delegation.ts
import * as path from 'path';
import * as fs from 'fs';
import {
  AgentCredential,
  BolyraConfig,
  DelegateeMerkleProof,
  DelegationResult,
  Proof,
} from './types';
import {
  BolyraError,
  CircuitArtifactNotFoundError,
  ProofGenerationError,
  ScopeEscalationError,
  VerificationError,
} from './errors';
import { eddsaSign, poseidon3, poseidon4 } from './utils';
import { proveGroth16, ProverBackend } from './prover';
import { loadSnarkjs } from './zk';

const DEFAULT_CIRCUIT_DIR =
  process.env.BOLYRA_CIRCUITS_DIR ?? path.join(__dirname, '../../circuits/build');

/** Delegation circuit MAX_DEPTH constant (matches circuits/src/Delegation.circom). */
const DELEGATION_MAX_DEPTH = 20;

/** snarkjs publicSignals layout for the Delegation circuit (outputs first):
 *    [0] newScopeCommitment
 *    [1] delegationNullifier
 *    [2] delegateeMerkleRoot
 *    [3] previousScopeCommitment
 *    [4] sessionNonce
 *    [5] currentTimestamp
 *  This MUST match IdentityRegistry.verifyDelegation()'s pubSignals layout.
 */
const PUBSIG_NEW_SCOPE = 0;
const PUBSIG_NULLIFIER = 1;
const PUBSIG_DELEGATEE_ROOT = 2;
const PUBSIG_PREV_SCOPE = 3;
const PUBSIG_SESSION_NONCE = 4;
const PUBSIG_CURRENT_TS = 5;

/** Build the single-leaf Merkle proof default (matches the conformance test pattern). */
function defaultMerkleProof(): DelegateeMerkleProof {
  return {
    length: 1,
    index: 0,
    siblings: new Array(DELEGATION_MAX_DEPTH).fill(0n),
  };
}

export interface DelegateInput {
  /** The delegating agent's credential (provides modelHash, operator pubkey, scope, expiry). */
  delegator: AgentCredential;
  /** Operator's EdDSA private key — used to sign the delegation token.
   *  Same key that minted the delegator credential. */
  delegatorOperatorPrivateKey: bigint | Buffer;
  /** Identity commitment of the recipient (leaf in agentTree). */
  delegateeCommitment: bigint;
  /** Narrowed scope being granted. Must be a subset of delegator.permissionBitmask
   *  under the cumulative-bit rules. Circuit enforces; SDK pre-checks for a clean error. */
  delegateeScope: bigint;
  /** Expiry being granted. Must be <= delegator.expiryTimestamp. */
  delegateeExpiry: bigint;
  /** Scope commitment from the prior chain link.
   *  For hop 1, this is the agent's scopeCommitment output from the handshake.
   *  For hop N+1, this is the previous delegation's newScopeCommitment. */
  previousScopeCommitment: bigint;
  /** Session nonce. Must match the nonce of the originating handshake. */
  sessionNonce: bigint;
  /** Unix-seconds timestamp the proof is bound to. Must be within MAX_CLOCK_SKEW
   *  of block.timestamp at on-chain verification time (300s in IdentityRegistry).
   *  Defaults to floor(Date.now() / 1000). */
  currentTimestamp?: bigint;
  /** Optional Merkle inclusion proof for the delegatee in agentTree.
   *  Defaults to the single-leaf pattern (sufficient for tests and demos). */
  delegateeMerkleProof?: DelegateeMerkleProof;
  /** Informational hop index (0-indexed). Not consumed by the circuit;
   *  echoed in DelegationResult for caller bookkeeping. */
  hopIndex?: number;
  config?: BolyraConfig;
  backend?: ProverBackend;
}

/**
 * Generate a Delegation proof.
 *
 * Produces a Groth16 proof matching the Delegation circuit. The caller submits
 * `{ proof, publicSignals }` to `IdentityRegistry.verifyDelegation()` along with
 * the session nonce.
 *
 * Scope narrowing is one-way: the circuit (and contract) reject any delegatee
 * scope that is not a subset of the delegator's, and any expiry past the
 * delegator's. The cumulative-bit invariants (bit 4 ⇒ 2+3, bit 3 ⇒ 2) are
 * enforced on the delegatee scope.
 *
 * @example
 * ```ts
 * const { proof, result } = await delegate({
 *   delegator: parentCredential,
 *   delegatorOperatorPrivateKey: operatorSecret,
 *   delegateeCommitment: childCredential.commitment,
 *   delegateeScope: 0b00000011n,                    // read + write (narrower)
 *   delegateeExpiry: parentCredential.expiryTimestamp - 3600n,
 *   previousScopeCommitment: handshake.scopeCommitment,
 *   sessionNonce: handshake.sessionNonce,
 * });
 * ```
 */
export async function delegate(
  input: DelegateInput,
): Promise<{ proof: Proof; result: DelegationResult }> {
  // Pre-flight scope check — gives a clean error before paying for proof generation.
  if (
    (input.delegateeScope & ~input.delegator.permissionBitmask) !== 0n
  ) {
    throw new ScopeEscalationError(
      input.delegator.permissionBitmask,
      input.delegateeScope,
    );
  }
  if (input.delegateeExpiry > input.delegator.expiryTimestamp) {
    throw new BolyraError(
      `Delegatee expiry (${input.delegateeExpiry}) exceeds delegator expiry (${input.delegator.expiryTimestamp}). Delegations may only narrow expiry, not extend it.`,
      'EXPIRY_ESCALATION',
      {
        delegateeExpiry: input.delegateeExpiry.toString(),
        delegatorExpiry: input.delegator.expiryTimestamp.toString(),
      },
    );
  }

  // Sanity check: the previousScopeCommitment passed in must equal the
  // identity-bound chain link Poseidon3(delegatorScope, delegatorCredCommitment,
  // delegatorExpiry). The circuit will assert this; we precheck for a clean error.
  // Ordered before artifact loading so CI environments without circuits still
  // surface CHAIN_LINK_MISMATCH instead of CIRCUIT_ARTIFACT_NOT_FOUND.
  const expectedPrev = await poseidon3(
    input.delegator.permissionBitmask,
    input.delegator.commitment,
    input.delegator.expiryTimestamp,
  );
  if (expectedPrev !== input.previousScopeCommitment) {
    throw new BolyraError(
      `previousScopeCommitment does not match the delegator's identity-bound chain link. Got ${input.previousScopeCommitment}, expected ${expectedPrev} (= Poseidon3(scope, credCommitment, expiry) for this delegator). For hop 1, pass the agent's scopeCommitment output from proveHandshake.`,
      'CHAIN_LINK_MISMATCH',
    );
  }

  const circuitDir = input.config?.circuitDir ?? DEFAULT_CIRCUIT_DIR;
  const backend = input.backend ?? 'auto';
  const currentTimestamp =
    input.currentTimestamp ?? BigInt(Math.floor(Date.now() / 1000));
  const merkleProof = input.delegateeMerkleProof ?? defaultMerkleProof();
  if (merkleProof.siblings.length !== DELEGATION_MAX_DEPTH) {
    throw new BolyraError(
      `Delegatee Merkle proof must have exactly ${DELEGATION_MAX_DEPTH} siblings (got ${merkleProof.siblings.length}).`,
      'INVALID_MERKLE_PROOF',
    );
  }

  const wasmPath = path.join(circuitDir, 'Delegation_js/Delegation.wasm');
  const zkeyPath = path.join(circuitDir, 'Delegation_final.zkey');
  if (!fs.existsSync(wasmPath)) {
    throw new CircuitArtifactNotFoundError(wasmPath, 'wasm');
  }
  if (!fs.existsSync(zkeyPath)) {
    throw new CircuitArtifactNotFoundError(zkeyPath, 'zkey');
  }

  // The delegator signs the delegation token, binding this delegation to a
  // specific recipient with specific scope+expiry.
  // Token = Poseidon4(previousScopeCommitment, delegateeCommitment, delegateeScope, delegateeExpiry).
  const tokenHash = await poseidon4(
    input.previousScopeCommitment,
    input.delegateeCommitment,
    input.delegateeScope,
    input.delegateeExpiry,
  );
  const sig = await eddsaSign(input.delegatorOperatorPrivateKey, tokenHash);

  const witnessInput: Record<string, unknown> = {
    delegatorScope: input.delegator.permissionBitmask.toString(),
    delegateeScope: input.delegateeScope.toString(),
    delegateeExpiry: input.delegateeExpiry.toString(),
    delegatorExpiry: input.delegator.expiryTimestamp.toString(),
    delegatorModelHash: input.delegator.modelHash.toString(),
    delegatorPubkeyAx: input.delegator.operatorPublicKey.x.toString(),
    delegatorPubkeyAy: input.delegator.operatorPublicKey.y.toString(),
    sigR8x: sig.R8.x.toString(),
    sigR8y: sig.R8.y.toString(),
    sigS: sig.S.toString(),
    delegatorCredCommitment: input.delegator.commitment.toString(),
    delegateeCredCommitment: input.delegateeCommitment.toString(),
    delegateeMerkleProofLength: merkleProof.length.toString(),
    delegateeMerkleProofIndex: merkleProof.index.toString(),
    delegateeMerkleProofSiblings: merkleProof.siblings.map((s) => s.toString()),
    previousScopeCommitment: input.previousScopeCommitment.toString(),
    sessionNonce: input.sessionNonce.toString(),
    currentTimestamp: currentTimestamp.toString(),
  };

  let proof: Proof;
  try {
    proof = await proveGroth16(witnessInput, wasmPath, zkeyPath, backend);
  } catch (err: any) {
    throw new ProofGenerationError('Delegation', err.message ?? String(err));
  }

  const result: DelegationResult = {
    newScopeCommitment: BigInt(proof.publicSignals[PUBSIG_NEW_SCOPE]),
    delegationNullifier: BigInt(proof.publicSignals[PUBSIG_NULLIFIER]),
    delegateeMerkleRoot: BigInt(proof.publicSignals[PUBSIG_DELEGATEE_ROOT]),
    hopIndex: input.hopIndex ?? 0,
  };

  return { proof, result };
}

/**
 * Verify a delegation proof off-chain (snarkjs Groth16 verify).
 *
 * For on-chain enforcement, submit `proof` and `proof.publicSignals` to
 * `IdentityRegistry.verifyDelegation(proof, pubSignals, sessionNonce)` — that
 * call additionally enforces chain state, hop count, expiry binding, and
 * nullifier replay. Off-chain verify here only confirms the proof itself is
 * mathematically valid and that the publicSignals match the expected chain link.
 */
export async function verifyDelegation(
  proof: Proof,
  previousScopeCommitment: bigint,
  sessionNonce: bigint,
  currentTimestamp: bigint,
  config?: BolyraConfig,
): Promise<DelegationResult> {
  if (!proof || !proof.proof || !Array.isArray(proof.publicSignals)) {
    throw new VerificationError(
      'Invalid Delegation proof structure: expected { proof: object, publicSignals: string[] }.',
    );
  }
  if (proof.publicSignals.length < 6) {
    throw new VerificationError(
      `Delegation proof has ${proof.publicSignals.length} public signals, expected 6.`,
    );
  }

  // Bind the public signals to the expected chain context before trusting the proof.
  if (BigInt(proof.publicSignals[PUBSIG_PREV_SCOPE]) !== previousScopeCommitment) {
    throw new VerificationError(
      `previousScopeCommitment mismatch: proof binds ${proof.publicSignals[PUBSIG_PREV_SCOPE]}, caller expected ${previousScopeCommitment}.`,
    );
  }
  if (BigInt(proof.publicSignals[PUBSIG_SESSION_NONCE]) !== sessionNonce) {
    throw new VerificationError(
      `sessionNonce mismatch: proof binds ${proof.publicSignals[PUBSIG_SESSION_NONCE]}, caller expected ${sessionNonce}.`,
    );
  }
  if (BigInt(proof.publicSignals[PUBSIG_CURRENT_TS]) !== currentTimestamp) {
    throw new VerificationError(
      `currentTimestamp mismatch: proof binds ${proof.publicSignals[PUBSIG_CURRENT_TS]}, caller expected ${currentTimestamp}.`,
    );
  }

  const circuitDir = config?.circuitDir ?? DEFAULT_CIRCUIT_DIR;
  const vkeyPath = path.join(circuitDir, 'Delegation_groth16_vkey.json');
  if (!fs.existsSync(vkeyPath)) {
    throw new CircuitArtifactNotFoundError(vkeyPath, 'vkey');
  }
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const vkey = require(vkeyPath);

  // ZK path begins here — snarkjs is loaded lazily so classical (Core)
  // callers of this module never pay the snarkjs module-load cost.
  const snarkjs = await loadSnarkjs();
  const valid = await snarkjs.groth16.verify(vkey, proof.publicSignals, proof.proof);
  if (!valid) {
    throw new VerificationError('Delegation proof failed Groth16 verification.');
  }

  return {
    newScopeCommitment: BigInt(proof.publicSignals[PUBSIG_NEW_SCOPE]),
    delegationNullifier: BigInt(proof.publicSignals[PUBSIG_NULLIFIER]),
    delegateeMerkleRoot: BigInt(proof.publicSignals[PUBSIG_DELEGATEE_ROOT]),
    hopIndex: 0,
  };
}

File: sdk-python/bolyra/delegation.py
"""Delegation proof generation and verification.

Thin subprocess bridge to ``@bolyra/sdk``'s ``delegate()`` and
``verifyDelegation()`` -- mirrors :mod:`bolyra.handshake`. ZK proving lives
in Node.js (snarkjs / rapidsnark); Python owns types, pre-flight validation,
and structured error mapping.

Scope narrowing is one-way: the underlying circuit and contract reject any
delegatee scope that is not a subset of the delegator's, and any expiry past
the delegator's. The cumulative-bit invariants (bit 4 implies 2+3, bit 3
implies 2) are enforced on the delegatee scope by the circuit.
"""

from __future__ import annotations

import json
import time

from bolyra._bridge import resolve_node_sdk, run_node_script
from bolyra.errors import BolyraError, ScopeEscalationError, VerificationError
from bolyra.types import (
    AgentCredential,
    BolyraConfig,
    DelegateeMerkleProof,
    DelegationResult,
    Proof,
)

# Delegation circuit MAX_DEPTH -- must match circuits/src/Delegation.circom.
DELEGATION_MAX_DEPTH = 20


def delegate(
    delegator: AgentCredential,
    delegator_operator_private_key: int,
    delegatee_commitment: int,
    delegatee_scope: int,
    delegatee_expiry: int,
    previous_scope_commitment: int,
    session_nonce: int,
    *,
    current_timestamp: int | None = None,
    delegatee_merkle_proof: DelegateeMerkleProof | None = None,
    hop_index: int = 0,
    config: BolyraConfig | None = None,
) -> tuple[Proof, DelegationResult]:
    """Generate a single-hop Delegation proof.

    Shells out to ``@bolyra/sdk``'s ``delegate()``. The Python side handles
    cheap pre-flight checks (scope escalation, expiry escalation, Merkle
    proof shape) so callers do not pay subprocess overhead for trivial
    rejections.

    Args:
        delegator: The delegating agent's credential.
        delegator_operator_private_key: EdDSA private key of the operator
            who minted ``delegator``. Used to sign the delegation token.
        delegatee_commitment: Identity commitment of the recipient.
        delegatee_scope: Narrowed permission bitmask being granted.
        delegatee_expiry: Expiry timestamp (must be <= delegator expiry).
        previous_scope_commitment: Chain link from the prior hop. For hop 1
            this is the handshake's ``scope_commitment``.
        session_nonce: Session nonce of the originating handshake.
        current_timestamp: Proof binding timestamp (defaults to now).
        delegatee_merkle_proof: Optional Merkle inclusion proof. Defaults to
            the single-leaf pattern (``length=1, index=0, siblings=[0]*20``).
        hop_index: Informational hop number; not consumed by the circuit.
        config: SDK configuration.

    Returns:
        Tuple of ``(Proof, DelegationResult)`` matching the TS SDK.

    Raises:
        ScopeEscalationError: If ``delegatee_scope`` is not a subset of
            ``delegator.permission_bitmask``.
        BolyraError: ``EXPIRY_ESCALATION`` if expiry is extended;
            ``INVALID_MERKLE_PROOF`` if siblings length is wrong.
        ProofGenerationError: If the Node.js bridge fails.
    """
    # Pre-flight in Python -- avoids the subprocess for obvious rejections.
    if (delegatee_scope & ~delegator.permission_bitmask) != 0:
        raise ScopeEscalationError(delegator.permission_bitmask, delegatee_scope)
    if delegatee_expiry > delegator.expiry_timestamp:
        raise BolyraError(
            f"Delegatee expiry ({delegatee_expiry}) exceeds delegator expiry "
            f"({delegator.expiry_timestamp}). Delegations may only narrow expiry.",
            "EXPIRY_ESCALATION",
            {
                "delegatee_expiry": str(delegatee_expiry),
                "delegator_expiry": str(delegator.expiry_timestamp),
            },
        )
    if delegatee_merkle_proof is not None and (
        len(delegatee_merkle_proof.siblings) != DELEGATION_MAX_DEPTH
    ):
        raise BolyraError(
            f"Delegatee Merkle proof must have exactly {DELEGATION_MAX_DEPTH} "
            f"siblings (got {len(delegatee_merkle_proof.siblings)}).",
            "INVALID_MERKLE_PROOF",
        )

    if current_timestamp is None:
        current_timestamp = int(time.time())

    merkle = delegatee_merkle_proof or DelegateeMerkleProof(
        length=1, index=0, siblings=[0] * DELEGATION_MAX_DEPTH
    )
    siblings_literal = "[" + ",".join(f"{s}n" for s in merkle.siblings) + "]"

    sdk_path = resolve_node_sdk(config)

    script = f"""
const {{ delegate }} = require('./dist/index.js');

async function main() {{
    const delegator = {{
        modelHash: {delegator.model_hash}n,
        operatorPublicKey: {{ x: {delegator.operator_public_key.x}n, y: {delegator.operator_public_key.y}n }},
        permissionBitmask: {delegator.permission_bitmask}n,
        expiryTimestamp: {delegator.expiry_timestamp}n,
        signature: {{
            R8: {{ x: {delegator.signature.r8.x}n, y: {delegator.signature.r8.y}n }},
            S: {delegator.signature.s}n,
        }},
        commitment: {delegator.commitment}n,
    }};
    const {{ proof, result }} = await delegate({{
        delegator,
        delegatorOperatorPrivateKey: {delegator_operator_private_key}n,
        delegateeCommitment: {delegatee_commitment}n,
        delegateeScope: {delegatee_scope}n,
        delegateeExpiry: {delegatee_expiry}n,
        previousScopeCommitment: {previous_scope_commitment}n,
        sessionNonce: {session_nonce}n,
        currentTimestamp: {current_timestamp}n,
        delegateeMerkleProof: {{
            length: {merkle.length},
            index: {merkle.index},
            siblings: {siblings_literal},
        }},
        hopIndex: {hop_index},
    }});
    const serialize = (obj) => JSON.stringify(obj, (_, v) =>
        typeof v === 'bigint' ? v.toString() : v
    );
    console.log(serialize({{ proof, result }}));
}}
main().catch(e => {{
    // Try to emit a structured error the Python side can recover.
    const payload = {{
        code: e.code || 'UNKNOWN_ERROR',
        message: e.message || String(e),
        details: e.details || {{}},
    }};
    console.error(JSON.stringify(payload));
    process.exit(1);
}});
"""

    data = run_node_script(script, sdk_path, op="delegation")
    proof = Proof(
        proof=data["proof"]["proof"],
        public_signals=data["proof"]["publicSignals"],
    )
    result_data = data["result"]
    result = DelegationResult(
        new_scope_commitment=int(result_data["newScopeCommitment"]),
        delegation_nullifier=int(result_data["delegationNullifier"]),
        delegatee_merkle_root=int(result_data["delegateeMerkleRoot"]),
        hop_index=int(result_data["hopIndex"]),
    )
    return proof, result


def verify_delegation(
    proof: Proof,
    previous_scope_commitment: int,
    session_nonce: int,
    current_timestamp: int,
    config: BolyraConfig | None = None,
) -> DelegationResult:
    """Verify a Delegation proof off-chain via snarkjs Groth16 verify.

    For on-chain enforcement, submit ``proof`` and ``proof.public_signals``
    to ``IdentityRegistry.verifyDelegation(proof, pubSignals, sessionNonce)``
    -- the contract additionally checks chain state, max-hops, expiry
    binding, and nullifier replay. Off-chain verify here only confirms the
    proof itself is mathematically valid and that the public signals match
    the expected chain context.

    Args:
        proof: The Delegation proof.
        previous_scope_commitment: Expected ``pubSignals[3]``.
        session_nonce: Expected ``pubSignals[4]``.
        current_timestamp: Expected ``pubSignals[5]``.
        config: SDK configuration.

    Returns:
        DelegationResult with ``hop_index=0`` (verification cannot infer it).

    Raises:
        VerificationError: If the public signals do not match or Groth16
            verification fails.
        ConfigurationError: If Node.js SDK is not found.
    """
    if not proof or not isinstance(proof.proof, dict) or not isinstance(
        proof.public_signals, list
    ):
        raise VerificationError(
            "Invalid Delegation proof structure: expected proof: dict, public_signals: list."
        )
    if len(proof.public_signals) < 6:
        raise VerificationError(
            f"Delegation proof has {len(proof.public_signals)} public signals, expected 6."
        )

    sdk_path = resolve_node_sdk(config)

    proof_json = json.dumps(proof.proof)
    signals_json = json.dumps(proof.public_signals)

    script = f"""
const {{ verifyDelegation }} = require('./dist/index.js');

async function main() {{
    const proof = {{
        proof: {proof_json},
        publicSignals: {signals_json},
    }};
    const result = await verifyDelegation(
        proof,
        {previous_scope_commitment}n,
        {session_nonce}n,
        {current_timestamp}n,
    );
    const serialize = (obj) => JSON.stringify(obj, (_, v) =>
        typeof v === 'bigint' ? v.toString() : v
    );
    console.log(serialize(result));
}}
main().catch(e => {{
    const payload = {{
        code: e.code || 'UNKNOWN_ERROR',
        message: e.message || String(e),
        details: e.details || {{}},
    }};
    console.error(JSON.stringify(payload));
    process.exit(1);
}});
"""

    data = run_node_script(script, sdk_path, op="delegation_verify")
    return DelegationResult(
        new_scope_commitment=int(data["newScopeCommitment"]),
        delegation_nullifier=int(data["delegationNullifier"]),
        delegatee_merkle_root=int(data["delegateeMerkleRoot"]),
        hop_index=0,
    )
